flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/GSASingleSourceShortestPathsExample.java

```diff
@@ -0,0 +1,216 @@
+	// --------------------------------------------------------------------------------------------
+
+	public static void main(String[] args) throws Exception {
+
+		if(!parseParameters(args)) {
+			return;
+		}
+
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+
+		DataSet<Vertex<Long, Double>> vertices = getVertexDataSet(env);
+		DataSet<Edge<Long, Double>> edges = getEdgeDataSet(env);
+
+		Graph<Long, Double, Double> graph = Graph.fromDataSet(vertices, edges, env);
+
+		// The path from src to trg through edge e costs src + e
+		// If the target's distance is 0 (the target is the actual source), return 0 instead
+		GatherFunction<Double, Double, Double> gather = new SingleSourceShortestPathGather();
+
+		// Return the smaller path length to minimize distance
+		SumFunction<Double, Double, Double> sum = new SingleSourceShortestPathSum();
+
+		// Iterate as long as the distance is updated
+		ApplyFunction<Double, Double, Double> apply = new SingleSourceShortestPathApply();
+
+		// Execute the GSA iteration
+		GatherSumApplyIteration<Long, Double, Double, Double> iteration = graph.createGatherSumApplyIteration(
+				gather, sum, apply, maxIterations);
+		Graph<Long, Double, Double> result = graph.mapVertices(new InitVerticesMapper<Long>(srcVertexId))
+				.runGatherSumApplyIteration(iteration);
+
+		// Extract the vertices as the result
+		DataSet<Vertex<Long, Double>> singleSourceShortestPaths = result.getVertices();
+
+		// emit result
+		if(fileOutput) {
+			singleSourceShortestPaths.writeAsCsv(outputPath, "\n", " ");
+		} else {
+			singleSourceShortestPaths.print();
+		}
+
+		env.execute("GSA Single Source Shortest Paths Example");
+	}
+
+	public static final class InitVerticesMapper<K extends Comparable<K> & Serializable>
+			implements MapFunction<Vertex<K, Double>, Double> {
+
+		private K srcVertexId;
+
+		public InitVerticesMapper(K srcId) {
+			this.srcVertexId = srcId;
+		}
+
+		public Double map(Vertex<K, Double> value) {
+			if (value.f0.equals(srcVertexId)) {
+				return 0.0;
+			} else {
+				return Double.POSITIVE_INFINITY;
+			}
+		}
+	}
+
+	// --------------------------------------------------------------------------------------------
+	//  Single Source Shortest Path UDFs
+	// --------------------------------------------------------------------------------------------
+
+	private static final class SingleSourceShortestPathGather
+			extends GatherFunction<Double, Double, Double> {
+		@Override
+		public Double gather(Triplet<Double, Double> triplet) {
+			return triplet.getSource() + triplet.getEdge();
+		}
+	};
+
+	private static final class SingleSourceShortestPathSum
+			extends SumFunction<Double, Double, Double> {
+		@Override
+		public Double sum(Double arg0, Double arg1) {
+			return Math.min(arg0, arg1);
+		}
+	};
+
+	private static final class SingleSourceShortestPathApply
+			extends ApplyFunction<Double, Double, Double> {
+		@Override
+		public void apply(Double summed, Double target) {
```

Comment:
I'd do it in the other example as well, yes, to make it easier to understand what it does.